Homework 7

Dragos-Andrei-75 opened the pull request:
Setup One to One relations between Librarian (owner) and Library entities.

Setup One to Many and Many to One relations between Book (owner) and Library entities.

Defined endpoints:
- create
    - handles requests to create a Librarian and a Library that the Librarian will be the owner of;
- addBook
    - handles requests to add a Book to the books list of a Library;
- removeBook
    - handles requests to remove a Book from the books list of a Library;
- other general endpoints;

Created classes:
- Librarians (all packages);
- Libraries (all packages);
- Books (all packages);

Refactored classes:
- UserService
    - added the code from the EmailService and VerificationService classes;
- UserController
    - added the code from the EmailController and VerificationController classes;

Deleted classes:
- EmailService;
- EmailController;
- VerificationService;
- VerificationController;

Defined enums:
- Category (category field of Book class);

Dragos-Andrei-75 committed:
Homework 7

Setup One to One relations between Librarian (owner) and Library entities.

Setup One to Many and Many to One relations between Book (owner) and Library entities.

Defined endpoints:
- create
    - handles requests to create a
      Librarian and a Library that the
      Librarian will be the owner of;
- addBook
    - handles requests to add a Book to
      the books list of a Library;
- removeBook
    - handles requests to remove a Book
      from the books list of a Library;
- other general endpoints;

Created classes:
- Librarians (all packages);
- Libraries (all packages);
- Books (all packages);

Refactored classes:
- UserService
    - added the code from the
      EmailService and
      VerificationService classes;
- UserController
    - added the code from the
      EmailController and
      VerificationController classes;

Deleted classes:
- EmailService;
- EmailController;
- VerificationService;
- VerificationController;

Defined enums:
- Category (category field of Book class);

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/Category.java:
```diff
@@ -0,0 +1,9 @@
+package com.example.Spring_Initializr_Bookstore;
```
m-as bucura, cand poti, sa redenumesti pachetul asta, sa scapam de acel "_".

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/Category.java:
```diff
@@ -0,0 +1,9 @@
+package com.example.Spring_Initializr_Bookstore;
+
+public enum Category {
```
Oare mai putem introduce inca un nume, pe langa "Category"? Maybe Book category? si poate cu pachetul sau separat

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,82 @@
+
+import java.util.List;
+
+@RestController()
```
Daca poti, sa scoti (), sa le folosim doar atunci cand avem ceva in interior

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,82 @@
+    @Autowired()
+    private LibraryService libraryService;
+
+    @PostMapping(path = "/create-book")
```
fara "create-book", doar Post

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,82 @@
+        return ResponseEntity.ok(BookMapper.book2BookDTO(createdBook));
+    }
+
+    @GetMapping(path = "/find-book/{bookID}")
```
fara find-book. analog si la restul, unde vorbim deja de "/books" si e clar din verb (get, put, post, delete) ce facem

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,82 @@
+
+    @GetMapping(path = "/list-all-books")
+    public ResponseEntity<?> listAll() {
+        List<Book> books = bookService.listAll();
```
as pune-o inline, mai jos

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,82 @@
+    @GetMapping(path = "/list-all-books")
+    public ResponseEntity<?> listAll() {
+        List<Book> books = bookService.listAll();
+        List<BookDTO> bookDTOS = books.stream().map(BookMapper::book2BookDTO).toList();
```
hai sa punem map() si toList() pe alta linie, mereu doar .stream() pe aceeasi linie, orice alta operatie sub, pt readability

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,82 @@
+
+    @PutMapping(path = "/update-book/{bookID}")
+    public ResponseEntity<?> update(@PathVariable(name = "bookID") Long bookID, @RequestBody() BookDTO bookDTO) {
+        Book bookToUpdate = bookService.findByID(bookID);
```
hmm... as merge in update() cu logica in care cauti cartea dupa id, adica in service sa faci find-ul and stuff, e mai safe ca sa fie si toata operatia in aceeasi tranzactie cu BD.

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,82 @@
+
+    @PutMapping(path = "/add/{bookID}/to/{libraryID}")
+    public ResponseEntity<?> addBook(@PathVariable(name = "bookID") Long bookID, @PathVariable(name = "libraryID") Long libraryID) {
+        Book bookToAdd = bookService.findByID(bookID);
```
analogi asta cu find, corectezi tu cand poti, please. logicile sa fie in service cat mai mult, cam asta ar fi ideea

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/controller/LibrarianController.java:
```diff
@@ -0,0 +1,59 @@
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+@RestController()
```
ca sa nu fac spam, nu uita deci de paranteze peste tot + create-etc sa nu mai fie in cale

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/controller/LibrarianController.java:
```diff
@@ -0,0 +1,59 @@
+
+    @PostMapping(path = "/verify-code")
+    public ResponseEntity<?> verifyAccount(@RequestParam() Long librarianID, @RequestParam() String code) {
+        Librarian librarian = librarianRepository.findById(librarianID).orElseThrow(() -> new EntityNotFoundException("User with ID " + librarianID + " not found."));
```
service logic, nu mai repet asta, tot asa, sa nu fac spam

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/controller/LibrarianController.java:
```diff
@@ -0,0 +1,59 @@
+        Librarian librarian = librarianRepository.findById(librarianID).orElseThrow(() -> new EntityNotFoundException("User with ID " + librarianID + " not found."));
+        String message = librarianService.checkVerificationCode(librarian, code);
+
+        return ResponseEntity.ok(LibrarianMapper.librarian2LibrarianDTO(librarian) + message);
```
fara "+ message", ori doar entitatea, ori 204 no content, please, ca altfel nu va putea fi deserializat json-ul intr-un obiect pe front-end.

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/controller/LibrarianController.java:
```diff
@@ -0,0 +1,59 @@
+
+        if (loginValidation) {
+            return ResponseEntity.noContent().build();
+        } else {
```
no need else, direct return

marcopacurariu3 replied:
hai la login sa returnam obiectul, altfel pe front-end dupa login nu stii ce id ai

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/controller/UserController.java:
```diff
@@ -64,6 +61,32 @@ public ResponseEntity<?> delete(@PathVariable(name = "userID") Long userID) {
         return ResponseEntity.ok().body("Deleted user with ID " + userID + ".");
     }
 
+    @PostMapping(path = "/send-email/{userID}")
```
do we need these ca si controller methods? in caz de retry send?

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/entities/User.java:
```diff
@@ -19,14 +20,15 @@ public class User {
 
     @Column(name = "YEAR_OF_BIRTH")
-    private Integer yearOfBirth;
+    private LocalDate yearOfBirth;
```
birthDate

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/entities/User.java:
```diff
@@ -40,12 +42,15 @@ public class User {
     private LocalDateTime verificationCodeGenerationTime;
 
+    @Column(name = "LOGGED_IN")
+    private Boolean loggedIn = false;
```
not sure if we need it. in viata reala iti expira sesiunea si nu o sa il puna nimeni pe false, decat daca am face un scheduler sau ceva, eventual. mai complicat. as scoate asta for now

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/entities/User.java:
```diff
@@ -87,11 +92,11 @@ public void setCountry(String country) {
 
     public void setYearOfBirth(Integer yearOfBirth) {
-        this.yearOfBirth = yearOfBirth;
+        this.yearOfBirth = LocalDate.ofYearDay(yearOfBirth, 1);
```
aici ziceam cu birthDate, ca hai sa lasam data reala, daca tot e LocalDate, altfel mergea Integer cum ai avut.

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/entitiesDTO/UserDTO.java:
```diff
@@ -55,11 +57,11 @@ public void setCountry(String country) {
     }
 
     public Integer getYearOfBirth() {
```
lasa si aici LocalDate atunci :)

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/mapper/BookMapper.java:
```diff
@@ -16,6 +27,17 @@ public static BookDTO book2BookDTO(Book book) {
+        bookDTO.setNumberOfPages(book.getNumberOfPages());
+        bookDTO.setReleaseDate(book.getReleaseDate());
+
+        if (book.getLibrary() != null) {
```
nice

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/mapper/LibrarianMapper.java:
```diff
@@ -4,19 +4,37 @@
+        librarianDTO.setPassword(librarian.getPassword());
+        librarianDTO.setVerifiedAccount(librarian.getVerifiedAccount());
+        librarian.setLoggedIn(librarianDTO.getLoggedIn());
+/*
```
remove if not needed

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/mapper/LibraryMapper.java:
```diff
@@ -8,6 +8,17 @@ public static Library libraryDTO2Library(LibraryDTO libraryDTO) {
+        library.setName(libraryDTO.getName());
+        library.setAddress(libraryDTO.getAddress());
+        library.setPhoneNumber(libraryDTO.getPhoneNumber());
+/*
```
remove sau uncomment

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/mapper/LibraryMapper.java:
```diff
@@ -8,6 +8,17 @@ public static Library libraryDTO2Library(LibraryDTO libraryDTO) {
+            library.setLibrarian(LibrarianMapper.librarianDTO2Librarian(libraryDTO.getLibrarianDTO()));
+        }
+*/
+        if (libraryDTO.getBookDTOS() != null && !libraryDTO.getBookDTOS().isEmpty()) {
```
e ok daca e empty, no need for that check

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/mapper/LibraryMapper.java:
```diff
@@ -8,6 +8,17 @@ public static Library libraryDTO2Library(LibraryDTO libraryDTO) {
+        }
+*/
+        if (libraryDTO.getBookDTOS() != null && !libraryDTO.getBookDTOS().isEmpty()) {
+            library.setBooks(libraryDTO.getBookDTOS().stream().map(BookMapper::bookDTO2Book).toList());
```
new line dupa stream

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/mapper/LibraryMapper.java:
```diff
@@ -16,6 +27,17 @@ public static LibraryDTO library2LibraryDTO(Library library) {
+        libraryDTO.setName(library.getName());
+        libraryDTO.setAddress(library.getAddress());
+        libraryDTO.setPhoneNumber(library.getPhoneNumber());
+/*
```
remove comment

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/mapper/LibraryMapper.java:
```diff
@@ -16,6 +27,17 @@ public static LibraryDTO library2LibraryDTO(Library library) {
+            libraryDTO.setLibrarianDTO(LibrarianMapper.librarian2LibrarianDTO(library.getLibrarian()));
+        }
+*/
+        if (library.getBooks() != null && !library.getBooks().isEmpty()) {
```
no need for is empty

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/service/BookService.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    public Book findByID(Long bookID) {
+        return bookRepository.findById(bookID).orElseThrow(() -> new EntityNotFoundException("Book with ID " + bookID + " not found."));
```
orElseThrow() pe new line

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/service/BookService.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    public Page<Book> listPaginated(Integer pageNumber, Integer numberOfElements) {
+        Pageable pageable = PageRequest.of(pageNumber, numberOfElements);
```
cam nimeni nu a facut asta:

Ar fi de asemenea frumos ca, daca acesti parametri lipsesc, sa cautati complet in baza de date, si doar daca ei sunt prezenti sa cautati paginat :)

dar las-o la final daca nu

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/service/LibrarianService.java:
```diff
@@ -0,0 +1,125 @@
+        return librarian;
+    }
+
+    public void send(String emailAddress, String subject, String text) {
```
asta mi se pare logica de EmailService

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/service/LibrarianService.java:
```diff
@@ -0,0 +1,125 @@
+
+            return "\nVerification code expired. Request a new verification code.";
+        } else if (!librarian.getVerificationCode().equals(code)) {
+            return "\nLibrarian account verification unsuccessful. Invalid code provided.";
```
arata bine, dar parca as arunca exceptie custom si as prinde-o in GlobalExceptionHandler si as arunca 400, sa returnezi string-uri e cam, asa, hidden sa zic. pur si simplu metoda "validateCode" si daca nu poate fi validat => exceptie

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/service/LibrarianService.java:
```diff
@@ -0,0 +1,125 @@
+    public Boolean login(Librarian librarian, String emailAddress, String password) {
+        String encryptedPassword = DigestUtils.md5DigestAsHex(password.getBytes(StandardCharsets.UTF_8));
+
+        if (!librarian.getEmail().equals(emailAddress) || !librarian.getPassword().equals(encryptedPassword))
```
exceptie aici. pe UI cum ar fi: primesti un mesaj de eroare, cam asa e si in RL in aplicatii

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/service/LibraryService.java:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    public Library findByID(Long libraryID) {
+        return libraryRepository.findById(libraryID).orElseThrow(() -> new EntityNotFoundException("Library with ID " + libraryID + " not found."));
```
new line

marcopacurariu3 commented on src/main/java/com/example/Spring_Initializr_Bookstore/service/UserService.java:
```diff
@@ -46,6 +57,65 @@ public void delete(User user) {
+        return user;
+    }
+
+    public void send(String recipient, String subject, String text) {
```
duplicated logic => EmailService

marcopacurariu3 commented:
Mersi mult ca ai formatat codul! Arata bine, par multe remarci, dar e doar polishing. Keep in touch, imi scrii daca e ceva!

Dragos-Andrei-75 committed:
Addressed All Issues from the Previous Review

Addressed all issues from the previous review made by Marco Păcurariu.

Added the Bookstore.postman_collection.json file into the resources folder.

Bookstore.postman_collection.json contains request for testing each endpoint that exist within the project.
